winlogbeat/eventlog/wineventlog.go

```diff
@@ -432,6 +426,47 @@ func (l *winEventLog) buildRecordFromXML(x []byte, recoveredErr error) Record {
 	return r
 }
 
+// winMetaCache retrieves and caches WinMeta tables by provider name.
+// It is a cut down version of the PublisherMetadataStore caching in wineventlog.Renderer.
+type winMetaCache struct {
+	mu     sync.RWMutex
+	cache  map[string]*winevent.WinMeta
+	logger *logp.Logger
+}
+
+func newWinMetaCache() winMetaCache {
+	return winMetaCache{cache: make(map[string]*winevent.WinMeta), logger: logp.L()}
+}
+
+func (c *winMetaCache) winMeta(provider string) *winevent.WinMeta {
```

Comment:
The behaviour of this function wrt cache invalidation matches that of the `wineventlog.Renderer` caching, that is that there is never a reason to invalidate. Is this safe? I imagine that it is, but want to make sure.

Reply:
From a size perspective I think it's going to safe in practice. There's a finite number of providers registered in the OS (< 1000) and usually it's only reading from a very small subset of them.

Reply:
And they never change meta data behaviour? This was my major concern rather than memory leaks.

Reply:
If the publishing software is updated then this could become stale. That could justify implementing a time-based expiration. I can't think of a reliable way to detecting that cache is stale or out of sync.